text/2022-05-12-multi-cluster-support.md

```diff
@@ -0,0 +1,335 @@
+# Multi Cluster Support
+
+## Summary
+
+As the multi-cluster injection is a huge feature, and related with the full
+lifecycle of Chaos Mesh, this RFC will be organized in a different way. It will
+describe the design through the Deploy -> Usage -> Authentication ->
+Architecture, and tell you the reason behind the design and other considerations
+in every separated section in a Q&A form.
+
+## Deploy
+
+A new cluster scope resource RemoteCluster will be used to install and manage
+the chaos-mesh installation on another cluster. This resource will be
+responsible to install, configure, upgrade and uninstall the cluster.
+
+```yaml
+apiVersion: chaos-mesh.org/v1alpha1
+kind: RemoteCluster
+metadata:
+ name: cluster-xxxxxxx
+spec:
+ namespace: "chaos-mesh"
+ kubeConfig:
+   secretRef:
+     name: “cluster-xxx-kubeconfig”
+   key: xxx
+ configOverride:
+   chaosDaemon:
+     runtime: containerd
+status:
+ currentVersion: "v2.1.0"
+ conditions:
+ - type: Installed
+   status: True
+ - type: Ready
+   status: True
+```
+
+The `Spec` defines the cluster which we want to install the chaos mesh, and the
+version of chaos mesh. It will install the same version of Chaos Mesh as the
+controller’s, which means if the version of controller is v2.2.0, it will
+install or upgrade the chaos mesh to v2.2.0 in the target cluster. It will only
+support the same minor version (e.g. parent cluster "v2.2.0" could control
+remote cluster "v2.2.1"). The `configOverride` allows to override the global
+configurations. The full configuration is a combination of the default
+configuration, the global configuration (provided through a constant configmap)
+and the override. We should provide an eliminated default configuration with the
+controller, as there are too many things not needed in the remote chaos mesh.
+
+The `Status` shows the current status of the target cluster. All of this field
+can be read from the target cluster through the helm package or read the pods’
+condition from the target cluster. The condition `Installed` is True iff the
+release has been installed in the target cluster and namespace. The condition
+`Ready` is True iff all pods are ready. The `currentVersion` in the status gives
+us a chance to automatically migrate the configuration.
+
+The total workflow can be described in the following graph:
+
+```
+User creates RemoteCluster -> RemoteCluster controller Reconcile
+
+User removes the RemoteCluster -> RemoteCluster controller Reconcile
+
+Pods in the target cluster/namespace changed -> RemoteCluster controller Reconcile (resources will be mapped to the `RemoteCluster`)
+
+RemoteCluster controller Reconcile
+-> If the resource is being deleted
+   -> Helm list the installed release
```

Comment:
Does this step run in the 'remote cluster'?

Reply:
Yes. I'll clarify it.